### ISSUE TITLE
[Question] How to set output audio device?

### ISSUE THREAD

**Issue by mirkancal** [ryanheise/audio_session#39]
Thanks to this package, we can get audio devices as a Future or Stream. I'm trying to cast audio to one of those devices. I couldn't figure out how I can do it. Is it possible with this package?

**Comment by ryanheise** [comment]
This plugin exposes the AVAudioSession API on iOS and the AudioManager API on Android (via the class `AndroidAudioManager`), which allow you to route audio. Since it's platform specific, I would recommend either looking up the platform docs for AVAudioSession/AudioManager (or you'll probably find answers on StackOverflow for how to use these two APIs to achieve the desired effect.)

**Comment by mirkancal** [comment]
I check couple SO posts, I read that it's not possible on iOS but it was an old answer. Anyway, I solve the iOS part by simple opening AirPlay. With this package [flutter_to_airplay](https://pub.dev/packages/flutter_to_airplay). Right now I'm looking to trigger to open this media view(native device picker view) on Android within the app. Since `audio_service` package register the audio within the OS, once I can trigger this native device picker, I'll be more than enough.

Back to the topic, for programmatically changing the audio route, I've checked this [SO](https://stackoverflow.com/questions/22770321/android-switching-audio-between-bluetooth-and-phone-speaker-is-inconsistent) post. I was looking for simple API call, like `audioManager.changeOutputDevice(deviceId: device.id)` but the example codes I see on the SO is more like stopping something first and then setting another thing, and I don't see anything related to the devices but calls like `setSpeakerphoneOn` or `setBluetoothScoOn`. I also see [just_audio](https://github.com/ryanheise/just_audio/issues/211) package has this casting feature on the roadmap, so if you can point me to where it's been developing at the moment, I'd love contribute.

**Comment by ryanheise** [comment]
Casting is another matter altogether, but you can take a look at [audio_cast](https://github.com/peterscodee/audio_cast) for the Android side of things.

As for the traditional audio devices, there are various answers on S/O:

* Android: https://stackoverflow.com/questions/22770321/android-switching-audio-between-bluetooth-and-phone-speaker-is-inconsistent
* iOS: https://stackoverflow.com/questions/52390659/avaudiosession-how-to-switch-between-speaker-and-headphones-output

**Comment by Vilmir** [comment]
I have not found any working option to list and change the audio output routes on iOS, apparently Apple does not want this to be modified programmatically. 
I have finally decided for ios to opt in for the option to invoke the OS audio route picker from my app. On iOS, it's [AVRoutePickerView](https://docs.flutter.dev/development/platform-integration/platform-views#on-the-dart-side), on Android [MediaRouteSelector](https://developer.android.com/reference/kotlin/androidx/mediarouter/media/MediaRouteSelector).
You can display native views in Flutter using the [official guide for native views](https://docs.flutter.dev/development/platform-integration/platform-views#on-the-dart-side).
It works pretty well after a minimal amount of work.

**Comment by ryanheise** [comment]
Some more info on iOS:

https://developer.apple.com/forums/thread/62954

It looks like you can programmatically set a Bluetooth input device, but you can't programmatically set a Bluetooth output device (unless that device is ALSO an input device, e.g. headset with a mic). For an output-only device, it needs to be done through a UI widget, so someone else would need to create a package for that to handle that use case.

**Comment by ryanheise** [comment]
On Android, the typically recommended solution involves deprecated methods. I think I may need to eventually implement the newer API (probably using the new JNIgen):

https://developer.android.com/reference/androidx/mediarouter/media/package-summary

**Comment by ryanheise** [comment]
Also suggestions from @Peng-Qian in #95 which I have folded into this issue. On the Android side, it seems to align with the usual S/O advice (and hence should probably be done with the newer `MediaRouter` API). Note that until these convenience methods are implemented, you can still achieve the goal with the current release of audio_session by manually invoking the same lower level methods as described in the solutions linked above (e.g. #95 ).

**Comment by ryanheise** [comment]
Here is an implementation of the suggested solutions (completely untested):

```dart
  final _androidAudioManager =
      !kIsWeb && Platform.isAndroid ? AndroidAudioManager() : null;
  final _avAudioSession = !kIsWeb && Platform.isIOS ? AVAudioSession() : null;

  Future<bool> switchToSpeaker() async {
    if (_androidAudioManager != null) {
      await _androidAudioManager!.setMode(AndroidAudioHardwareMode.normal);
      await _androidAudioManager!.stopBluetoothSco();
      await _androidAudioManager!.setBluetoothScoOn(false);
      await _androidAudioManager!.setSpeakerphoneOn(true);
    } else if (_avAudioSession != null) {
      await _avAudioSession!
          .overrideOutputAudioPort(AVAudioSessionPortOverride.speaker);
    }
    return true;
  }

  Future<bool> switchToReceiver() async {
    if (_androidAudioManager != null) {
      _androidAudioManager!.setMode(AndroidAudioHardwareMode.inCommunication);
      _androidAudioManager!.stopBluetoothSco();
      _androidAudioManager!.setBluetoothScoOn(false);
      _androidAudioManager!.setSpeakerphoneOn(false);
      return true;
    } else if (_avAudioSession != null) {
      return await _switchToAnyIosPortIn({AVAudioSessionPort.builtInMic});
    }
    return false;
  }

  Future<bool> switchToHeadphones() async {
    if (_androidAudioManager != null) {
      _androidAudioManager!.setMode(AndroidAudioHardwareMode.inCommunication);
      _androidAudioManager!.stopBluetoothSco();
      _androidAudioManager!.setBluetoothScoOn(false);
      _androidAudioManager!.setSpeakerphoneOn(false);
      return true;
    } else if (_avAudioSession != null) {
      return await _switchToAnyIosPortIn({AVAudioSessionPort.headsetMic});
    }
    return true;
  }

  Future<bool> switchToBluetooth() async {
    if (_androidAudioManager != null) {
      await _androidAudioManager!
          .setMode(AndroidAudioHardwareMode.inCommunication);
      await _androidAudioManager!.startBluetoothSco();
      await _androidAudioManager!.setBluetoothScoOn(true);
      return true;
    } else if (_avAudioSession != null) {
      return await _switchToAnyIosPortIn({
        AVAudioSessionPort.bluetoothLe,
        AVAudioSessionPort.bluetoothHfp,
        AVAudioSessionPort.bluetoothA2dp,
      });
    }
    return false;
  }

  Future<bool> _switchToAnyIosPortIn(Set<AVAudioSessionPort> ports) async {
    if ((await _avAudioSession!.currentRoute)
        .outputs
        .any((r) => ports.contains(r.portType))) {
      return true;
    }
    for (var input in await _avAudioSession!.availableInputs) {
      if (ports.contains(input.portType)) {
        await _avAudioSession!.setPreferredInput(input);
      }
    }
    return false;
  }
```

On Android, I did not provide the `setWiredHeadsetOn` because it is not only deprecated, but actually does nothing on newer versions of Android.

**Comment by Peng-Qian** [comment]
Hi @ryanheise,

Thank you for extending the invitation. I've been quite tied up recently, but I have indeed managed to test and improve the aforementioned code. During the process, I also discovered that on iOS, the audio session possesses remarkable audio routing capabilities. It's noteworthy that this is the only package I've found that performs this task well.

I'd be more than happy to submit a pull request, which I'll likely get around to next week. However, it's worth mentioning that due to my non-native developer background, you may need to carry out some refactoring pertaining to the file structure and location.

**Comment by mikemcelligott** [comment]
Just a heads up that I'm getting OSStatus error -50 when attempting to use the code above (though it does compile and looks like it's trying to do the right thing).  When I print out the list of devices on my iPhone I only get (id, name, type.name) the below (even though spotify is currently playing on a bluetooth speaker).  I'm guessing the app has to break out to a native chooser to connect a bluetooth / airplay output to the current app or something, as suggested above.

flutter: iPhone Microphone, Built-In Microphone, builtInMic
flutter: Speaker, Speaker, builtInSpeaker

**Comment by mvn-hangtran-dn** [comment]
@ryanheise I used `AndroidAudioHardwareMode.inCommunication`, but the text was not speaked fully sometimes on Android 13

**Comment by ryanheise** [comment]
Are you saying that it is routing to the correct hardware route but just that the audio is partial? I'm not sure if there is anything I can do in audio_session to address that.

**Comment by mvn-hangtran-dn** [comment]
@ryanheise Kindly testing functions of audio_session with new android versions(Android 13, 14) if you can. Many thanks!

**Comment by ryanheise** [comment]
I still don't understand what you mean since you didn't answer my question.

But aside from that, I don't actually see there is anything I can do in audio_session to address that. After all, the API you are using simply passes through to the operating system. The `setMode` method just passes through to the operating system, so that method does whatever the operating system does. I can't change what the operating system does.

**Comment by mvn-hangtran-dn** [comment]
@ryanheise 
https://github.com/ryanheise/audio_session/issues/39#issuecomment-1972472162
-> Yes, you got my problem correctly. It routes to the correct hardware route but just that the audio is partial.
Seem like `setMode` doesn't finish immediately, it need to a delay time.

**Comment by ryanheise** [comment]
Also don't forget to await the call to `setMode()`:

```dart
await setMode(...);
await Future.delayed(Duration(seconds: 1)); // experiment as needed
await playSound();
```

**Comment by mvn-hangtran-dn** [comment]
`await setMode(...);` -> I always use await
`await Future.delayed(Duration(seconds: 1));` -> I tried and the problem was fixed but the behaviour of my app is not good. For other android versions, I don't need it. I don't know why I need use this delay on Android 13.

**Comment by ryanheise** [comment]
This is the behaviour of the operating system. As suspected above, there isn't anything I can do to change the way the operating system works.

**Comment by zombie6888** [comment]
this works for me on ios:

```dart 
Future<bool> switchToSpeaker() async {
    if (_androidAudioManager != null) {
      await _androidAudioManager.setMode(AndroidAudioHardwareMode.normal);
      await _androidAudioManager.stopBluetoothSco();
      await _androidAudioManager.setBluetoothScoOn(false);
      await _androidAudioManager.setSpeakerphoneOn(true);
    } else if (_avAudioSession != null) {
      await _avAudioSession
          .overrideOutputAudioPort(AVAudioSessionPortOverride.speaker);
    }
    return true;
  }

  Future<bool> switchToReceiver() async {
    if (_androidAudioManager != null) {
      await _androidAudioManager
          .setMode(AndroidAudioHardwareMode.inCommunication);
      await _androidAudioManager.stopBluetoothSco();
      await _androidAudioManager.setBluetoothScoOn(false);
      await _androidAudioManager.setSpeakerphoneOn(false);
      return true;
    } else if (_avAudioSession != null) {
      await _avAudioSession
          .overrideOutputAudioPort(AVAudioSessionPortOverride.none);
      return _switchToAnyIosPortIn({AVAudioSessionPort.builtInMic});
    }
    return false;
  } 

  Future<bool> switchToBluetooth() async {
    if (_androidAudioManager != null) {
      await _androidAudioManager
          .setMode(AndroidAudioHardwareMode.inCommunication);
      await _androidAudioManager.startBluetoothSco();
      await _androidAudioManager.setBluetoothScoOn(true);
      return true;
    } else if (_avAudioSession != null) {
      return _switchToAnyIosPortIn({
        AVAudioSessionPort.bluetoothLe,
        AVAudioSessionPort.bluetoothHfp,
        AVAudioSessionPort.bluetoothA2dp,
      });
    }
    return false;
  }

  Future<bool> _switchToAnyIosPortIn(Set<AVAudioSessionPort> ports) async {   
    for (final input in await _avAudioSession!.availableInputs) {
      if (ports.contains(input.portType)) {
        await _avAudioSession.setPreferredInput(input);
      }
    }
    return false;
  }
```

tested with this config.

```dart
 await _avAudioSession?.setCategory(
        AVAudioSessionCategory.playAndRecord,
        AVAudioSessionCategoryOptions.allowBluetooth,
        AVAudioSessionMode.spokenAudio,);
```

We are planning to use this plugin for video calls. We are tried to use flutter-webrtc which is also includes audio output changing functionality but this works much better and i found this plugin more flexible. There are a lot of non-obvious things in the ios avaudiosession api. For example, in some cases to change the output device you need to set preffered input device etc. 

I will try Android part later and come back with feedback. Can we expect this functionality to be added to plugin if everything works?

**Comment by ryanheise** [comment]
Awesome! I appreciate your experimentation here, and yes I would definitely consider something like this being integrated into the main `AudioSession` class. On the Android side, there are new APIs for audio devices that I'll eventually need to get around to looking at, but for now we will have to go with the existing APIs in AudioManager (unless we run into a problem that can only be solved by implementing the new APIs).

**Comment by zombie6888** [comment]
The Android part also works. There are few cases that i haven't tested. These are different configs for ios, and usb/jack headphones, because i don't have such devices. I will let you know if I encounter any issues

**Comment by bradrf** [comment]
Has anyone tried [the solution](https://github.com/ryanheise/audio_session/issues/39#issuecomment-2222348470) provided by @zombie6888 and @ryanheise (thank you!) and found that, on Android, it does indeed work, but something ends up changing the audio output behind the scenes? For example, I start out with "speaker" output, then change to "receiver", but after a short period (10 or 20 seconds, maybe?) it switches back to "speaker". Ideas on what might be going on?

**Comment by zombie6888** [comment]
> Has anyone tried [the solution](https://github.com/ryanheise/audio_session/issues/39#issuecomment-2222348470) provided by @zombie6888 and @ryanheise (thank you!) and found that, on Android, it does indeed work, but something ends up changing the audio output behind the scenes? For example, I start out with "speaker" output, then change to "receiver", but after a short period (10 or 20 seconds, maybe?) it switches back to "speaker". Ideas on what might be going on?

Do you have this permission in your AndroidManifest.xml?
```xml
<uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
```

**Comment by bradrf** [comment]
> Do you have this permission in your AndroidManifest.xml?

No, we don't have that permission, and, when added, doesn't change the behavior. We do have `RECORD_AUDIO` requested, though, so I suspect that might include the ability to modify audio settings, too, but don't know for sure.

FWIW, a workaround that seems to "fix" the issue is to always invoke the desired function to select output just before playing an audio file.

**Comment by cddqssc** [comment]
@zombie6888 
(iOS) Could you please let me know if it's possible to record audio using the built-in microphone and play it back through a Bluetooth headset? I've tried numerous methods but couldn't achieve this.
Input and output seem to be limited to the same device.

**Comment by zombie6888** [comment]
> (iOS) Could you please let me know if it's possible to record audio using the built-in microphone and play it back through a Bluetooth headset? I've tried numerous methods but couldn't achieve this.
> Input and output seem to be limited to the same device.

I am not an expert but i think it's impossible do at the same time with the same audiosession. When you switch to built-in mic, the output device will also be changed to built-in earpiece. As far as i know this is due to legacy avaudiosession api. When I consulted with iOS developer, he mentioned that there are more advanced and  modern APIs for working with audio in iOS but i didn't try them.

**Comment by dganzella** [comment]
Is it possible to select the output device on web?

**Comment by Konvich** [comment]
Hello, and thank you for working on this plugin!

I also found the code above to work, and I wanted to ask whether there are plans to include it into the `AudioSession` class, @ryanheise?

**Comment by ryanheise** [comment]
> I also found the code above to work, and I wanted to ask whether there are plans to include it into the `AudioSession` class, [@ryanheise](https://github.com/ryanheise)?

I would be happy to merge a pull request if someone wants to contribute based on the discussion above, and once people can test it and confirm it working. Perhaps, @zombie6888 ?

This would have to be based on the branch submitted for PR #155 since that is going to be the next release (because it changes the default conditional compilation setting, it is also a breaking change).

**Comment by dganzella** [comment]
> > Is it possible to select the output device on web?
> 
> Web works a bit differently, since the output device needs to be set as a property of the media player object itself. For this reason, an API was recently added to just_audio called `setWebSinkId` which might be what you're after.

Yeah it was me who added it hahahaha. You already forgot my face :(

**Comment by ryanheise** [comment]
Oh, oops! I remembered debugChicken but forgot dganzella.

**Comment by zombie6888** [comment]
> I also found the code above to work, and I wanted to ask whether there are plans to include it into the `AudioSession` class

There is no need to add this functionality to the AudioSession class. Instead, we created a standalone service and passed it an AudioSession instance. I can share our solution or contribute to the current repository. This service is quite simple and reliable, it has been used in our major b2c app for almost a year.  But i need some time for PR and maybe add some example.

**Comment by ryanheise** [comment]
> for PR it would probably be better to move this functionality there

Yes, I think it would be preferable to have such convenience methods in the AudioSession class since it was originally intended as a convenience class to abstract common functionality to both Android and iOS. Although I would only want to add APIs here that are solid and permanent, since if I have to remove or change them in the future, it would cause a breaking change which I'd like to avoid due to the impact on the ecosystem that depends on specific versions of this package.

**Comment by zombie6888** [comment]
Done. @ryanheise  Please, look at: https://github.com/ryanheise/audio_session/pull/156

**Comment by hoalarious** [comment]
Thanks so much for this. Bump for merge please!

On my device I had to set

      await _androidAudioManager.setMode(AndroidAudioHardwareMode.inCommunication);

For the output to switch to speaker correctly while a bluetooth headset was connected.

    Future<bool> switchToSpeaker() async {
      try {
        if (_androidAudioManager != null) {
          print("Switching to speaker on Android");
          await _androidAudioManager.setMode(AndroidAudioHardwareMode.inCommunication);
          await _androidAudioManager.stopBluetoothSco();
          await _androidAudioManager.setBluetoothScoOn(false);
          await _androidAudioManager.setSpeakerphoneOn(true);
        } else if (_avAudioSession != null) {
          print("Switching to speaker on iOS");
          await _avAudioSession.overrideOutputAudioPort(AVAudioSessionPortOverride.speaker);
        }
        return true;
      } catch (e) {
        print("Error switching to speaker: $e");
        return false;
      }
    }

**Comment by zombie6888** [comment]
> On my device I had to set
> 
> ```
>   await _androidAudioManager.setMode(AndroidAudioHardwareMode.inCommunication);
> ```

That makes sense. I am not sure if this will work for every cases (we are using communication mode by default), but at least, it works properly in example. So i've added this change to MR. Thank you!
